packages/@okta/vuepress-site/docs/guides/oie-embedded-sdk-use-case-custom-pwd-recovery-mfa/main/java/siw.md

```diff
@@ -0,0 +1,122 @@
+
+### 2: Handle the OTP and state parameters
+
+Create a callback handler method that takes the `otp` and `state` parameters in the query string and passes them as session parameters to a page that contains the Sign-In Widget. Before calling the callback handler method, retrieve the current `IDXClientContext` object from the browser session and check that `state` has a value.
+
+If `state` is `null` or the browser context object is `null` (because the user clicked the magic link from a different browser), the widget instructs advise the user to return to the original tab in the browser where they requested a password reset and enter the OTP to proceed.
+
+```java
+@GetMapping("/magic-link/callback")
+public ModelAndView handleMagicLinkCallback(
+   HttpServletRequest request,
+   @RequestParam(name = "state") String state,
+   @RequestParam(name = "otp") String otp,
+   HttpSession session) throws MalformedURLException
+{
+  if (session.getAttribute(IDX_CLIENT_CONTEXT) == null) {
+      try {
+         idxClientContext = idxAuthenticationWrapper.getClientContext();
+      }
+      catch (ProcessingException e) {
+         // return error details
+      }
+      session.setAttribute(IDX_CLIENT_CONTEXT, idxClientContext);
+  }
+
+  if (idxClientContext == null) {
+     // return error details
+  }
+
+  if (state == null) {
+      return new ModelAndView("redirect:" + oktaOAuth2Properties.getRedirectUri());
+  }
+
+```
+
+If both values are valid, pass them (along with all the values required by the widget) to the page hosting the widget in your application.
+
+```java
+  String issuer = oktaOAuth2Properties.getIssuer();
+  // the widget needs the base url, just grab the root of the issuer
+  String orgUrl = new URL(new URL(issuer), "/").toString();
+
+  ModelAndView mav = new ModelAndView("login");
+  mav.addObject(STATE, state);
+  mav.addObject(OTP, otp);
+  mav.addObject(SCOPES, oktaOAuth2Properties.getScopes());
+  mav.addObject(OKTA_BASE_URL, orgUrl);
+  mav.addObject(OKTA_CLIENT_ID, oktaOAuth2Properties.getClientId());
+  mav.addObject(INTERACTION_HANDLE, idxClientContext.getInteractionHandle());
+  mav.addObject(CODE_VERIFIER, idxClientContext.getCodeVerifier());
+  mav.addObject(CODE_CHALLENGE, idxClientContext.getCodeChallenge());
+  mav.addObject(CODE_CHALLENGE_METHOD, CODE_CHALLENGE_METHOD_VALUE);
+  mav.addObject(REDIRECT_URI,
+    request.getScheme() + "://" + request.getServerName() + ":" +
+    request.getServerPort() + request.getContextPath() +
+    "/authorization-code/callback");
+  mav.addObject(ISSUER_URI, issuer);
+  session.setAttribute(CODE_VERIFIER, idxClientContext.getCodeVerifier());
+  return mav;
+}
+```
+
+You can pass the values to the widget using the widget's `config` object.
```

Comment:
How about something simpler:

> Pass the values using the widget's `config` object.

I say this because "to the widget" is implied by using both "pass" and "using the widget's"

Reply:
Absolutely - I remember struggling to phrase this well.